### PR TITLE
* Refine the laundry-washer-control server implementation and add laundry-washer-controls-delegate-impl

### PR DESCRIPTION
**Pull request by crlonxp**
* Add laundry-washer-controls-delegate-impl

### CODE REVIEW

File: examples/all-clusters-app/all-clusters-common/src/laundry-washer-controls-delegate-impl.cpp [hunk 1]
```diff
@@ -0,0 +1,42 @@
+#include <app/clusters/laundry-washer-controls-server/laundry-washer-controls-server.h>
+#include <app/util/config.h>
+#include <laundry-washer-controls-delegate-impl.h>
+
+using namespace std;
```
Comment (andy31415) [inline examples/all-clusters-app/all-clusters-common/src/laundry-washer-controls-delegate-impl.cpp:5]: where is std used?

In other files we generally do like you did in the header:

```
namespace chip {
namespace app {
namespace Clusters {
namespace LaundryWasherControls {
...
```

any reason for the many using here instead?

File: src/app/clusters/laundry-washer-controls-server/laundry-washer-controls-delegate.h [hunk 2]
```diff
@@ -35,20 +35,20 @@ class Delegate
...
+     *        the length to the length of the copied data.
+     * @return CHIP_ERROR_PROVIDER_LIST_EXHAUSTED if the index is out of range for the list of spin speeds.
      */
     virtual CHIP_ERROR GetSpinSpeedAtIndex(size_t index, MutableCharSpan & spinSpeed) = 0;
 
     /**
-     * Get the list of supported rinses list.
-     * Fills in the provided rinses  at index `index` if there is one,
-     * or returns CHIP_ERROR_PROVIDER_LIST_EXHAUSTED if the index is out of range for the list of rinses.
+     * Get the supported rinses string at the given index in the list.
```
Comment (bzbarsky-apple) [inline src/app/clusters/laundry-washer-controls-server/laundry-washer-controls-delegate.h:47]:
```suggestion
     * Get the supported rinses value at the given index in the list.
```
It's not a string!

File: src/app/clusters/laundry-washer-controls-server/laundry-washer-controls-delegate.h [hunk 3]
```diff
@@ -35,20 +35,20 @@ class Delegate
...
      * @param index The index of the supported rinses with 0 representing the first one.
-     * @param supportedRinse  The supported rinse is filled.
+     * @param supportedRinse The MutableCharSpan to copy the string data into.  On success, the callee must update
```
Comment (bzbarsky-apple) [inline src/app/clusters/laundry-washer-controls-server/laundry-washer-controls-delegate.h:49]: No, its not a MutableCharSpan.

File: src/app/clusters/laundry-washer-controls-server/laundry-washer-controls-delegate.h [hunk 4]
```diff
@@ -35,20 +35,20 @@ class Delegate
...
+     *        the length to the length of the copied data.
+     * @return CHIP_ERROR_PROVIDER_LIST_EXHAUSTED if the index is out of range for the list of spin speeds.
```
Comment (bzbarsky-apple) [inline src/app/clusters/laundry-washer-controls-server/laundry-washer-controls-delegate.h:51]: What do spin speeds have to do with this function?

File: examples/all-clusters-app/all-clusters-common/include/laundry-washer-controls-delegate-impl.h [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+#include <app/util/config.h>
+#include <cstring>
+
+namespace chip {
+namespace app {
+namespace Clusters {
+namespace LaundryWasherControls {
+
+/**
+ * The application delegate to statically defines the options.
```
Comment (bzbarsky-apple) [inline examples/all-clusters-app/all-clusters-common/include/laundry-washer-controls-delegate-impl.h:33]:
```suggestion
 * The application delegate to statically define the options.
```

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]

File: examples/all-clusters-app/all-clusters-common/src/laundry-washer-controls-delegate-impl.cpp (hunk 1 above)
Comment (crlonxp) [reply examples/all-clusters-app/all-clusters-common/src/laundry-washer-controls-delegate-impl.cpp:5]: They are the ones I forget removing. Will remove them.

File: src/app/clusters/laundry-washer-controls-server/laundry-washer-controls-delegate.h (hunk 3 above)
Comment (crlonxp) [reply src/app/clusters/laundry-washer-controls-server/laundry-washer-controls-delegate.h:49]: I will fix it.

File: src/app/clusters/laundry-washer-controls-server/laundry-washer-controls-delegate.h (hunk 4 above)
Comment (crlonxp) [reply src/app/clusters/laundry-washer-controls-server/laundry-washer-controls-delegate.h:51]: It's just to return the value at the index. I will change the description

Comment (crlonxp) [reply src/app/clusters/laundry-washer-controls-server/laundry-washer-controls-delegate.h]: spin-speed is not related with this function. I will change it

**Review by bzbarsky-apple — APPROVED** [review]

**Comment by crlonxp** [comment]
Are there anything we need to address for this PR? If not, could anybody help us to merge it? Thank you!

**Comment by bzbarsky-apple** [comment]
@crlonxp Needs a non-Apple review.   Please ping people on Slack?

File: examples/all-clusters-app/all-clusters-common/include/laundry-washer-controls-delegate-impl.h [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+    static const CharSpan spinSpeedsNameOptions[];
+    static const NumberOfRinsesEnum supportRinsesOptions[];
+    static LaundryWasherControlDelegate instance;
+
+public:
+    CHIP_ERROR GetSpinSpeedAtIndex(size_t index, MutableCharSpan & spinSpeed);
+    CHIP_ERROR GetSupportedRinseAtIndex(size_t index, NumberOfRinsesEnum & supportedRinse);
+
+    ~LaundryWasherControlDelegate(){};
+    LaundryWasherControlDelegate(){};
```
Comment (arkq) [inline examples/all-clusters-app/all-clusters-common/include/laundry-washer-controls-delegate-impl.h:47]: Why not:
```suggestion
    LaundryWasherControlDelegate() = default;
    ~LaundryWasherControlDelegate() = default;
```
?

**Review by arkq — APPROVED** [review]

**Review by andy31415 — APPROVED** [review]

**Review by cliffamzn — APPROVED** [review]

**Pull request merged**